Sources/MapboxCoreNavigation/NavigationRouter.swift

```diff
@@ -0,0 +1,179 @@
+                                                 completionHandler: @escaping Directions.RouteCompletionHandler) -> RequestId {
+        return doRequest(options: options) { [weak self] (result: Result<RouteResponse, DirectionsError>) in
+            guard let self = self else { return }
+            let session = (options: options as DirectionsOptions,
+                           credentials: self.settings.directions.credentials)
+            completionHandler(session, result)
+        }
+    }
+    
+    @discardableResult public func requestRoutes(options: MatchOptions,
+                                                 completionHandler: @escaping Directions.MatchCompletionHandler) -> RequestId {
+        return doRequest(options: options) { (result: Result<MapMatchingResponse, DirectionsError>) in
+            let session = (options: options as DirectionsOptions,
+                           credentials: self.settings.directions.credentials)
+            completionHandler(session, result)
+        }
+    }
+    
+    @discardableResult public func refreshRoute(indexedRouteResponse: IndexedRouteResponse,
+                                                fromLegAtIndex startLegIndex: UInt32 = 0,
+                                                completionHandler: @escaping Directions.RouteRefreshCompletionHandler) -> RequestId {
+        guard case let .route(routeOptions) = indexedRouteResponse.routeResponse.options,
+              let route = indexedRouteResponse.selectedRoute,
+              let responseIdentifier = indexedRouteResponse.routeResponse.identifier else {
+            preconditionFailure("Invalid route data passed for refreshing.")
+        }
+        
+        let encoder = JSONEncoder()
+        encoder.userInfo[.options] = routeOptions
+        
+        let routeIndex = UInt32(indexedRouteResponse.routeIndex)
+        
+        guard let routeData = try? encoder.encode(route),
+              let routeJSONString = String(data: routeData, encoding: .utf8) else {
+            preconditionFailure("Could not serialize route data for refreshing.")
+        }
+        
+        var requestId: RequestId!
+        requestId = router.getRouteRefresh(for: RouteRefreshOptions(requestId: responseIdentifier,
+                                                                    routeIndex: routeIndex,
+                                                                    legIndex: startLegIndex),
+                                           route: routeJSONString) { [weak self] result, _ in
+            guard let self = self else { return }
+            // change to route deserialization
+            do {
+                let json = result.value as? String
+                guard let data = json?.data(using: .utf8) else {
+                    DispatchQueue.main.async {
+                        completionHandler(self.settings.directions.credentials, .failure(.noData))
+                        self.finish(request: requestId)
+                    }
```

Comment:
Calling this completion can be refactored with simple local function with signature like:
```swift
func finish(self: NavigationRouter, result: Result<RouteRefreshResponse, DirectionsError>)
```

it will reduce duplication and the likehood of introducing a bug during future evolution.
Passing `self` will make it sure that no unexpected retain cycles introduced. 

Alternative would be to have this function at the type level as the same is needed in the route request above.

Reply:
Done. Thanks for suggestion!